// @storybook/react-native/dist/manager/provider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("core-js/modules/es7.symbol.async-iterator");

require("core-js/modules/es6.symbol");

require("core-js/modules/es6.object.define-property");

require("core-js/modules/es6.object.create");

require("core-js/modules/es6.object.set-prototype-of");

var _react = _interopRequireDefault(require("react"));

var _ui = require("@storybook/ui");

var _channelWebsocket = _interopRequireDefault(require("@storybook/channel-websocket"));

var _addons = _interopRequireDefault(require("@storybook/addons"));

var _coreEvents = _interopRequireDefault(require("@storybook/core-events"));

var _uuid = _interopRequireDefault(require("uuid"));

var _PreviewHelp = _interopRequireDefault(require("./components/PreviewHelp"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var ReactProvider =
/*#__PURE__*/
function (_Provider) {
  _inherits(ReactProvider, _Provider);

  function ReactProvider(_ref) {
    var _this;

    var domain = _ref.url,
        options = _ref.options;

    _classCallCheck(this, ReactProvider);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(ReactProvider).call(this));
    _this.options = options;
    _this.selection = null;

    try {
      _this.channel = _addons.default.getChannel();
    } catch (err) {
      _this.channel = undefined;
    }

    var secured = options.secured,
        host = options.host,
        port = options.port;
    var websocketType = secured ? 'wss' : 'ws';
    var url = "".concat(websocketType, "://").concat(domain);

    if (options.manualId) {
      _this.pairedId = (0, _uuid.default)();
      url += "/pairedId=".concat(_this.pairedId);
    }

    if (!_this.channel) {
      _this.channel = (0, _channelWebsocket.default)({
        url: url
      });

      _addons.default.setChannel(_this.channel);

      _this.channel.emit(_coreEvents.default.CHANNEL_CREATED, {
        pairedId: _this.pairedId,
        secured: secured,
        host: host,
        port: port
      });
    }

    return _this;
  }

  _createClass(ReactProvider, [{
    key: "getPanels",
    value: function getPanels() {
      return _addons.default.getPanels();
    }
  }, {
    key: "renderPreview",
    value: function renderPreview(kind, story) {
      this.selection = {
        kind: kind,
        story: story
      };
      this.channel.emit(_coreEvents.default.SET_CURRENT_STORY, {
        kind: kind,
        story: story
      });

      var renderPreview = _addons.default.getPreview();

      var innerPreview = renderPreview ? renderPreview(kind, story) : null;
      return innerPreview || _react.default.createElement(_PreviewHelp.default, null);
    }
  }, {
    key: "handleAPI",
    value: function handleAPI(api) {
      var _this2 = this;

      api.onStory(function (kind, story) {
        _this2.selection = {
          kind: kind,
          story: story
        };

        _this2.channel.emit(_coreEvents.default.SET_CURRENT_STORY, _this2.selection);
      });
      this.channel.on(_coreEvents.default.SELECT_STORY, function (_ref2) {
        var kind = _ref2.kind,
            story = _ref2.story;
        api.selectStory(kind, story);
      });
      this.channel.on(_coreEvents.default.SET_STORIES, function (data) {
        api.setStories(data.stories);
      });
      this.channel.on(_coreEvents.default.GET_CURRENT_STORY, function () {
        _this2.channel.emit(_coreEvents.default.SET_CURRENT_STORY, _this2.selection);
      });
      this.channel.emit(_coreEvents.default.GET_STORIES);

      _addons.default.loadAddons(api);
    }
  }]);

  return ReactProvider;
}(_ui.Provider);

exports.default = ReactProvider;
