// @storybook/channel-websocket/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createChannel;
exports.WebsocketTransport = void 0;

require("core-js/modules/es6.object.define-property");

require("core-js/modules/web.dom.iterable");

require("core-js/modules/es6.array.for-each");

var _global = require("global");

var _channels = _interopRequireDefault(require("@storybook/channels"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var WebsocketTransport =
/*#__PURE__*/
function () {
  function WebsocketTransport(_ref) {
    var url = _ref.url,
        onError = _ref.onError;

    _classCallCheck(this, WebsocketTransport);

    this._socket = null;
    this._buffer = [];
    this._handler = null;
    this._isReady = false;

    this._connect(url, onError);
  }

  _createClass(WebsocketTransport, [{
    key: "setHandler",
    value: function setHandler(handler) {
      this._handler = handler;
    }
  }, {
    key: "send",
    value: function send(event) {
      if (!this._isReady) {
        this._sendLater(event);
      } else {
        this._sendNow(event);
      }
    }
  }, {
    key: "_sendLater",
    value: function _sendLater(event) {
      this._buffer.push(event);
    }
  }, {
    key: "_sendNow",
    value: function _sendNow(event) {
      var data = JSON.stringify(event);

      this._socket.send(data);
    }
  }, {
    key: "_flush",
    value: function _flush() {
      var _this = this;

      var buffer = this._buffer;
      this._buffer = [];
      buffer.forEach(function (event) {
        return _this.send(event);
      });
    }
  }, {
    key: "_connect",
    value: function _connect(url, onError) {
      var _this2 = this;

      this._socket = new _global.WebSocket(url);

      this._socket.onopen = function () {
        _this2._isReady = true;

        _this2._flush();
      };

      this._socket.onmessage = function (e) {
        var event = JSON.parse(e.data);

        _this2._handler(event);
      };

      this._socket.onerror = function (e) {
        if (onError) {
          onError(e);
        }
      };
    }
  }]);

  return WebsocketTransport;
}();

exports.WebsocketTransport = WebsocketTransport;

function createChannel(_ref2) {
  var url = _ref2.url,
      async = _ref2.async,
      onError = _ref2.onError;
  var transport = new WebsocketTransport({
    url: url,
    onError: onError
  });
  return new _channels.default({
    transport: transport,
    async: async
  });
}
